### PR TITLE
Stabilize rfc 1506 - Clarified ADT Kinds

### LINKED ISSUE #35626: Tracking issue for "clarified ADT kinds" (RFC 1506)
Tracking issue for rust-lang/rfcs#1506.

**Status:**

**Still unstable, need a PR**

- [ ]  Numeric fields in struct expressions/patterns S { 0: a, 1: b } are not yet stabilized.
    - We've decided to stabilize the existing support, but we're still in need of a PR. If you're interested in contributing, [directions for stabilizing this feature can be found here](https://github.com/rust-lang/rust/issues/35626#issuecomment-287145810).

**Stabilized by PR https://github.com/rust-lang/rust/pull/36868:**

- [x] Lifted restrictions on tuple structs/variants are stabilized, i.e. S{..} can be used with any structs and empty tuple structs are permitted without feature gate.

### PR DESCRIPTION
**Pull request by matthewjasper**
Closes #35626

Documentation:

- [x] Reference rust-lang-nursery/reference#37
- [ ] Book?
- [ ] Rust by example?

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Thanks! We'll try to get someone to review your PR soon! r? @pnkfelix

**Comment by pnkfelix** [comment]
Wait, can I merge this before the docs land?

**Comment by pnkfelix** [comment]
(It certainly looks fine; this is more a process issue.)

**Comment by arielb1** [comment]
I think we should be waiting for the docs to land to avoid this being split across 1.17/1.18. but r=me

**Comment by arielb1** [comment]
Still waiting for the docs to land on reference.

**Comment by arielb1** [comment]
Still waiting for the docs to land on reference. But there's a light at the end of the tunnel:

> Sorry about this! Still working out how to best do these kinds of things. I will get to it early next week; right now RustFest is going on.

**Comment by carols10cents** [comment]
@arielb1 was that @steveklabnik you were quoting there? What's the current status of this?

**Comment by brson** [comment]
This is a real easy feature to land for 1.19, just need to resolve [this docs pr](https://github.com/rust-lang-nursery/reference/pull/37) in the next two weeks.

**Comment by steveklabnik** [comment]
@brson I have wanted to get on it but haven't; IMHO, since we have the PR, this can land. Generally it's

1. send in docs pr
2. land the pr
3. land the docs pr

so we don't churn feature flags in the docs

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -1291,6 +1287,11 @@ impl<'a> Visitor<'a> for PostExpansionVisitor<'a> {
             ast::ExprKind::InPlace(..) => {
                 gate_feature_post!(&self, placement_in_syntax, e.span, EXPLAIN_PLACEMENT_IN);
             }
+<<<<<<< HEAD
```
Comment (Gordon-F) [inline src/libsyntax/feature_gate.rs]: Unfinished conflict?

**Comment by petrochenkov** [comment]
Given @steveklabnik's comment https://github.com/rust-lang/rust/pull/41145#issuecomment-303242014
@bors r+

**Pull request merged**